rclrs/src/timer.rs

```diff
@@ -0,0 +1,416 @@
+        let ret = unsafe { rcl_timer_reset(&mut *timer) };
+
+        // rcl_timer_reset should only error if incorrect arguments are given
+        // or something isn't initialised, both of which we control in this function.
+        debug_assert_eq!(RclReturnCode::try_from(ret).unwrap(), RclReturnCode::Ok);
+    }
+
+    /// Internal function to check the timer is still valid and set the last call time in rcl.
+    /// Returns [`RclReturnCode::TimerCanceled`] as an error if the timer has already been canceled.
+    fn call_rcl(&self) -> Result<(), RclrsError> {
+        let mut timer = self.handle.lock();
+        // SAFETY: Safe if the timer is initialized, which is guaranteed by the constructor.
+        unsafe {
+            rcl_timer_call(&mut *timer).ok()?;
+        }
+        Ok(())
+    }
+}
+
+impl TimerBase for Timer {
+    fn handle(&self) -> &TimerHandle {
+        &self.handle
+    }
+
+    fn execute(&mut self) -> Result<(), RclrsError> {
+        // Timer still needs to be called within RCL, even though we handle the callback ourselves.
+        self.call_rcl()?;
+
+        let callback = self.callback.clone();
+        (*callback.lock().unwrap())(self);
+
+        Ok(())
+    }
+}
+
+// Timer.rs does very little other than call rcl functions.
+// To keep these tests easy to maintain, most of them just check the rcl functions
+// can be called without returning errors.
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use crate::{Clock, Context};
+
+    use super::Timer;
+
+    fn new_timer() -> Timer {
+        let context = Context::new([]).unwrap();
+
+        // This is technically a wall clock, but we have a period of 0 so it won't slow down unit testing.
+        let clock = Clock::system();
+
+        let timer = Timer::new(&context, clock, Duration::from_secs(0), |_| {});
+
+        timer.expect("Timer::new should not return an error")
+    }
+
+    #[test]
+    fn creation() {
+        let _ = new_timer();
+    }
+
+    #[test]
+    fn is_ready() {
+        let timer = new_timer();
+
+        // Calling is_ready will trigger the debug_assert check on the rcl return value.
+        timer.is_ready();
+    }
+
+    #[test]
+    fn time_until_next_call() {
+        let timer = new_timer();
+
+        timer
+            .time_until_next_call()
+            .expect("Calling Timer::time_until_next_call on a non-canceled timer should not error");
+    }
+
+    #[test]
+    fn time_since_last_call() {
+        let timer = new_timer();
+
+        // Calling time_since_last_call will trigger the debug_assert check on the rcl return value.
+        timer.time_since_last_call();
+    }
+
+    #[test]
+    fn update_period() {
+        let timer = new_timer();
+
+        let new_period = Duration::from_millis(100);
+
+        // Calling set_period will trigger the debug_assert check on the rcl return value.
+        timer.set_period(new_period.clone());
```

Comment:
As per: https://github.com/geoff-imdex/ros2_rust/actions/runs/11564633206/job/32190192780#step:11:395

```suggestion
        timer.set_period(new_period);
```

Reply:
Good catch, fixed.